### PR TITLE
Implement `pub(restricted)` privacy (RFC 1422)

### PR DESCRIPTION
**Pull request by jseyfried**
This implements `pub(restricted)` privacy from https://github.com/rust-lang/rfcs/pull/1422 (cc #32409) behind a feature gate.

`pub(restricted)` paths currently cannot use re-exported modules both for simplicity of implementation and for future compatibility with https://github.com/rust-lang/rfcs/pull/1560 (cc #31783).

r? @nikomatsakis

### CODE REVIEW

**Comment by jseyfried** [comment]
I didn't allow `pub(restricted)` for tuple struct fields, as suggested [here](https://github.com/rust-lang/rust/issues/32409#issuecomment-205553822).

Also `pub(restricted)` imports cannot be used in non-lexical scopes (like private imports) so that `pub(self)` and the absence of a visibility modifier are equivalent (as suggested [here](https://github.com/rust-lang/rfcs/pull/1422#issuecomment-195525227)). Once RFC 1560 is accepted and #32213 lands, all imports will be usable in non-lexical scopes (provided that they are visible enough, of course).

**Comment by jseyfried** [comment]
cc @petrochenkov @pnkfelix

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -949,6 +928,9 @@ pub struct NameBinding<'a> {
     modifiers: DefModifiers,
     kind: NameBindingKind<'a>,
     span: Option<Span>,
+    // Enum variants are always considered `PUBLIC`, this is needed for `use Enum::Variant`
+    // or `use Enum::*` to work on private enums.
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: This comment is no longer true after https://github.com/rust-lang/rust/pull/32875/commits/d1ef3561305cb43071c34fbbfad8ecc649996889?

File: src/test/compile-fail/private-in-public-warn.rs [hunk 2]
```diff
@@ -287,5 +288,15 @@ mod aliases_params {
     type Result<T> = ::std::result::Result<T, Priv>;
 }
 
+mod foo {
+    struct Priv;
+    pub mod bar {
+        use super::Priv;
+        pub(super) fn f(_: Priv) {} // ok
+        pub(crate) fn g(_: Priv) {} //~ WARN private
```
Comment (petrochenkov) [inline src/test/compile-fail/private-in-public-warn.rs]: Heh, it's unfortunate that completely new errors are reported as warnings just because they are absent in the Old Error Set.

Comment (petrochenkov) [inline src/test/compile-fail/private-in-public-warn.rs]: Maybe private-in-public visitor one more temporary flag to report everything as errors if `pub(restricted)` is involved (not necessarily precisely).

File: src/librustc_resolve/lib.rs (hunk 1 above)
Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: Indeed, fixed.

File: src/test/compile-fail/private-in-public-warn.rs (hunk 2 above)
Comment (jseyfried) [inline src/test/compile-fail/private-in-public-warn.rs]: > report everything as errors if pub(restricted) is involved

I like that idea.

File: src/libsyntax/visit.rs [hunk 3]
```diff
@@ -619,6 +621,7 @@ pub fn walk_struct_def<'v, V: Visitor<'v>>(visitor: &mut V,
 
 pub fn walk_struct_field<'v, V: Visitor<'v>>(visitor: &mut V,
                                              struct_field: &'v StructField) {
+    visitor.visit_vis(&struct_field.vis);
     walk_opt_ident(visitor, struct_field.span, struct_field.ident);
     visitor.visit_ty(&struct_field.ty);
     walk_list!(visitor, visit_attribute, &struct_field.attrs);
```
Comment (petrochenkov) [inline src/libsyntax/visit.rs]: Foreign items seem to be missing.

Comment (petrochenkov) [inline src/libsyntax/visit.rs]: And HIR visibilities in librustc\hir\intravisit.rs are not visited.

File: src/librustc/ty/mod.rs [hunk 4]
```diff
@@ -287,6 +287,13 @@ impl Visibility {
     pub fn from_hir(visibility: &hir::Visibility, id: NodeId, tcx: &TyCtxt) -> Self {
         match *visibility {
             hir::Public => Visibility::Public,
+            hir::Visibility::Crate => Visibility::Restricted(ast::CRATE_NODE_ID),
+            hir::Visibility::Restricted { id, .. } => match tcx.def_map.borrow().get(&id) {
+                Some(resolution) => Visibility::Restricted({
+                    tcx.map.as_local_node_id(resolution.base_def.def_id()).unwrap()
+                }),
+                None => Visibility::Public,
```
Comment (petrochenkov) [inline src/librustc/ty/mod.rs]: Does compilation continue when a path in `pub(path)` is unresolved? Is `Visibility::Public` used here to avoid spurious privacy errors?

File: src/librustc_resolve/build_reduced_graph.rs [hunk 5]
```diff
@@ -434,32 +428,27 @@ impl<'b, 'tcx:'b> Resolver<'b, 'tcx> {
...
-        let is_public = xcdef.vis == ty::Visibility::Public || parent.is_trait();
-
-        let mut modifiers = DefModifiers::empty();
-        if is_public {
-            modifiers = modifiers | DefModifiers::PUBLIC;
-        }
-        if parent.is_normal() {
-            modifiers = modifiers | DefModifiers::IMPORTABLE;
-        }
+        let vis = if parent.is_trait() { ty::Visibility::Public } else { xcdef.vis };
```
Comment (petrochenkov) [inline src/librustc_resolve/build_reduced_graph.rs]: Hmm, is this trick still need with `pub(restricted)`? I guess `xcdef.vis` should just work now, no?

File: src/librustc_resolve/lib.rs [hunk 6]
```diff
@@ -894,7 +895,7 @@ impl<'a> ModuleS<'a> {
 
     fn is_normal(&self) -> bool {
         match self.def {
-            Some(Def::Mod(_)) | Some(Def::ForeignMod(_)) => true,
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: I have no idea what `is_normal` means and why `Def::ForeignMod` was removed from here :D 
Could you explain?

File: src/librustc/ty/mod.rs (hunk 4 above)
Comment (jseyfried) [inline src/librustc/ty/mod.rs]: Yes (to both questions), I'll add a comment explaining this.

File: src/librustc_resolve/build_reduced_graph.rs (hunk 5 above)
Comment (jseyfried) [inline src/librustc_resolve/build_reduced_graph.rs]: Good point, I believe `xcdef.vis` will just work now.

File: src/librustc_resolve/lib.rs [hunk 7]
```diff
@@ -982,7 +977,19 @@ impl<'a> NameBinding<'a> {
     }
 
     fn is_public(&self) -> bool {
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: I'd get rid of this function (preferably), or at least rename it to `is_pseudo_public`

File: src/libsyntax/visit.rs (hunk 3 above)
Comment (jseyfried) [inline src/libsyntax/visit.rs]: > Foreign items seem to be missing.

Good catch, I'll add them.

> And HIR visibilities in librustc\hir\intravisit.rs are not visited.

There's no need to visit visibilities in the HIR right now -- do you think I should a `visit_vis` method in `intravisit` anyway?

File: src/librustc_resolve/build_reduced_graph.rs (hunk 5 above)
Comment (jseyfried) [inline src/librustc_resolve/build_reduced_graph.rs]: `xcdef.vis` is still `PrivateExternal` for trait methods. I believe this is because visibilities of trait methods are not encoded in the metadata and the decoder considers items without encoded visibilities to be `PrivateExternal`.

Ideally, the decoder would give trait methods the same visibility as the trait itself, but that is orthogonal to this PR.

File: src/libsyntax/parse/parser.rs [hunk 8]
```diff
@@ -560,7 +561,9 @@ impl<'a> Parser<'a> {
     }
 
     pub fn parse_ident(&mut self) -> PResult<'a, ast::Ident> {
-        self.check_strict_keywords();
+        if !self.restrictions.contains(Restrictions::ALLOW_MODULE_PATHS) {
+            self.check_strict_keywords();
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Lol.
I'm pretty sure, naked `self` and `super` should *always* be parseable as paths/idents and resolved later as usual.
There are even a couple of issued about this - https://github.com/rust-lang/rust/issues/29036, https://github.com/rust-lang/rust/issues/27899.
But not this time, I guess.

File: src/libsyntax/visit.rs (hunk 3 above)
Comment (petrochenkov) [inline src/libsyntax/visit.rs]: >do you think I should a visit_vis method in intravisit

Yes, default visitor functions just traverse the tree exhaustively, providing, well, correct defaults.
As an example, overridden `visit_path`s will not visit paths in visibilities if `intravisit` doesn't traverse visibilities by default.

File: src/librustc_resolve/lib.rs (hunk 6 above)
Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: `is_normal` means something that `self` or `super` resolve to -- it is used when resolving path prefixes.
`ForeignMod`s should never have been normal, but happened to not make a difference before this PR. Now, `ForeignMod` can't be normal since otherwise a `pub(self)` item in a foreign module would be restricted to the foreign module instead of its containing module.

File: src/libsyntax/visit.rs (hunk 3 above)
Comment (jseyfried) [inline src/libsyntax/visit.rs]: Ok, will do.

File: src/librustc_resolve/lib.rs (hunk 7 above)
Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: Will do.

File: src/libsyntax/parse/parser.rs (hunk 8 above)
Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: > `self` and `super` should always be parseable as paths/idents

Agreed. I'll change that and fix the issues you referenced in another PR.

**Comment by jseyfried** [comment]
I rebased and addressed @petrochenkov's comments.

File: src/libsyntax/visit.rs (hunk 3 above)
Comment (petrochenkov) [inline src/libsyntax/visit.rs]: I've just noticed, default folders (`fold.rs`) have the same problems as visitors.
AST folders don't fold *all* visibilities and HIR folders don't fold visibilities at all.

Comment (jseyfried) [inline src/libsyntax/visit.rs]: Good point, I'll fix this.

File: src/librustc_resolve/build_reduced_graph.rs (hunk 5 above)
Comment (nikomatsakis) [inline src/librustc_resolve/build_reduced_graph.rs]: @jseyfried 

> Ideally, the decoder would give trait methods the same visibility as the trait itself, but that is orthogonal to this PR.

I think we will propose at some point to be able to add `pub` declarations to trait methods, in which case I guess we'd have to encode the visibility. But I agree: orthogonal.

File: src/test/compile-fail/pub-restricted-2.rs [hunk 9]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
```
Comment (nikomatsakis) [inline src/test/compile-fail/pub-restricted-2.rs]: Meta note: Let's make a subdirectory for these tests.

Perhaps `src/test/compile-fail/privacy`? If we want to further designate tests aimed at the `pub(restricted)` RFC, then perhaps `src/test/compile-fail/privacy/restricted`?

**Comment by nikomatsakis** [comment]
This looks great. I suggest we move the privacy tests to a subdirectory. I want to start giving more structure to our `compile-fail` directory on new PRs. :) I'd also like to start maintaining a list that aims to show what is being tested (and, ideally, where). To that end, I drew up a list of stuff that I think we ought to test from this RFC and went through and checked off what you were already testing. The good news is that you got almost all of it! I'm going to write up my matrix here, maybe you can add tests for the remaining few items. In some cases, the tests are probably a bit redundant: I was just trying to think -- without reference to the impl -- of the various scenarios that might arise and try to be somewhat exhaustive about it. 

So, here is my matrix. I've put an `X` for things that I thought you were testing, a `-` for cases that don't really apply, and left blank things that seem untested. I also added in an `O` in a few spots for the things I'd really *like* to see tests for. The rest I think you can add tests for if you think it makes sense.

```
pub(crate)                                              | X | X |
pub(super)                                              | X | X |
pub(other::path)                                        | X |   |
pub(bad::path)                                          | X | - |
pub(not::ancestor)                                      | X | - |
method calls                                            | X | X |
field access                                            | X | X |
struct literal where fields are private                 | X |   |
field lookup ignores fields with pub(foo)               | O | O |
method lookup ignores fields with pub(foo)              | O | O |
associated item notation (UFCS)                         |   |   |
  trait method (in scope, from parent)                  |   |   |
  inherent method (in scope, from parent)               |   |   |
  where trait is out of scope                           | O | O |
  where method is out of scope (inherent methods only)  | O | O |
reference of item in use statement                      | X | X |
reference of item in expression                         | X | X |
reference of item in type definition                    |   |   |
reference of trait in an impl                           |   |   |
reference of self-type in a trait impl                  |   |   |
reference of self-type in an inherent impl              |   |   |
feature-gate required                                   | X | - |
private type in public API:                             |   |   |
  fn arg                                                | X |   |
  other cases?                                          |   |   |
```

Eventually I'd like to have a system for tagging tests with what they test so we can easily construct matrices like this automatically to get some idea of test coverage. But we're not there yet.

r=me with suitable tests added.

File: src/libsyntax/parse/parser.rs (hunk 8 above)
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Actually, I'm going to fix this myself if you don't mind.

Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: Excellent, I hadn't started anything yet.

**Comment by jseyfried** [comment]
@nikomatsakis 
What is the X-axis? I'm guessing something like [ error emitted | no error emitted ].
Other than that, sounds good.

File: src/libsyntax/visit.rs (hunk 3 above)
Comment (jseyfried) [inline src/libsyntax/visit.rs]: Done

**Comment by nikomatsakis** [comment]
@jseyfried oh, I forgot to label those :) it was "intra-crate" and "cross-crate"

**Comment by nikomatsakis** [comment]
@jseyfried if you want me to review add'l tests, let me know, but if you feel like you've satisfied all the interesting points in that matrix, then r=me

**Comment by jseyfried** [comment]
Ok, makes sense. I'll add some more tests and r=you

**Comment by petrochenkov** [comment]
A note: visibilities do not expect type parameters in paths, but use `parse_path` for parsing, so paths with type parameters can be supplied to them with macros `$p: path => pub($p)`, so visibilities need to be sanity-checked after expansion. I'll fix this as a part of https://github.com/rust-lang/rust/pull/32875#discussion_r59641163.

**Comment by jseyfried** [comment]
@petrochenkov interesting, sounds good.
